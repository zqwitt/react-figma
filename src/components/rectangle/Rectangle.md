## Rectangle

Wrapper for the Figma [Rectangle](https://www.figma.com/plugin-docs/api/RectangleNode/).

#### Props

| Prop       | Type     | Default | Note                                              |
| ---------- | -------- | ------- | ------------------------------------------------- |
| `name`     | `String` |         | The name to be displayed in the Figma Layers List |
| `top` | `Number` |  |  |
| `left` | `Number` |  |  |
| `width` | `Number` |  |  |
| `height` | `Number` |  |  |
| `style`    | [`Style`](/docs/styling.md)   |  | Not all props |
| `onSelectionEnter` | `Function` |  | Selection enter event callback  |
| `onSelectionLeave` | `Function` |  | Selection leave event callback  |

Also, most of the [RectangleNode](https://www.figma.com/plugin-docs/api/RectangleNode/) fields are supported as props.

#### Example

```javascript
<Rectangle
    height={100}
    width={200}
/>
```
