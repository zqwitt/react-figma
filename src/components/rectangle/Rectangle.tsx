import * as React from 'react';
import { DefaultShapeProps, BorderProps, CornerProps, StyleOf } from '../../types';
import {
    LayoutStyleProperties,
    transformLayoutStyleProperties
} from '../../styleTransformers/transformLayoutStyleProperties';
import {
    GeometryStyleProperties,
    transformGeometryStyleProperties
} from '../../styleTransformers/transformGeometryStyleProperties';
import { useYogaLayout } from '../../hooks/useYogaLayout';
import { useFillsPreprocessor } from '../../hooks/useFillsPreprocessor';
import {
    BorderStyleProperties,
    transformBorderStyleProperties
} from '../../styleTransformers/transformBorderProperties';
import { transformBlendProperties, BlendStyleProperties } from '../../styleTransformers/transformBlendProperties';
import { StyleSheet } from '../..';

export interface RectangleProps extends DefaultShapeProps, CornerProps, BorderProps {
    style?: StyleOf<LayoutStyleProperties & GeometryStyleProperties & BorderStyleProperties & BlendStyleProperties>;
    children?: undefined;
}

export const Rectangle: React.FC<RectangleProps> = props => {
    const nodeRef = React.useRef();

    const style = StyleSheet.flatten(props.style);

    const rectangleProps = {
        ...transformLayoutStyleProperties(style),
        ...transformGeometryStyleProperties('fills', style),
        ...transformBorderStyleProperties(style),
        ...transformBlendProperties(style),
        ...props
    };
    const fills = useFillsPreprocessor(rectangleProps);
    const yogaProps = useYogaLayout({ nodeRef, ...rectangleProps });

    // @ts-ignore
    return <rectangle {...rectangleProps} {...yogaProps} {...(fills && { fills })} innerRef={nodeRef} />;
};
