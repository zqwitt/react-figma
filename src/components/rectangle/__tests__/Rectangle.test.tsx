import * as React from 'react';
const renderer = require('react-test-renderer');
import { Rectangle } from '../Rectangle';
import { createFigma } from 'figma-api-stub';

describe('<Rectangle />', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
    });

    it('Rectangle without props', () => {
        const tree = renderer.create(<Rectangle />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Rectangle with layout style properties', () => {
        const tree = renderer
            .create(
                <Rectangle
                    style={{
                        width: 100,
                        height: 200,
                        top: 50,
                        left: 20
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Rectangle with color background', () => {
        const tree = renderer
            .create(
                <Rectangle
                    style={{
                        backgroundColor: '#1bff00'
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Rectangle with image background', () => {
        const tree = renderer
            .create(
                <Rectangle
                    style={{
                        backgroundColor: '#ff8b29',
                        backgroundSize: 'cover',
                        backgroundImage: './image.png'
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Rectangle with border props', () => {
        const tree = renderer
            .create(
                <Rectangle
                    style={{
                        width: 200,
                        height: 100,
                        borderWidth: 10,
                        borderColor: '#ff8b29'
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
