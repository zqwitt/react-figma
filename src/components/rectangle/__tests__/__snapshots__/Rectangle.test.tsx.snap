// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rectangle /> Rectangle with border props 1`] = `
<rectangle
  height={100}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  strokeWeight={10}
  strokes={
    Array [
      Object {
        "color": Object {
          "b": 0.1607843137254902,
          "g": 0.5450980392156862,
          "r": 1,
        },
        "type": "SOLID",
      },
    ]
  }
  style={
    Object {
      "borderColor": "#ff8b29",
      "borderWidth": 10,
      "height": 100,
      "width": 200,
    }
  }
  width={200}
/>
`;

exports[`<Rectangle /> Rectangle with color background 1`] = `
<rectangle
  fills={
    Array [
      Object {
        "color": Object {
          "b": 0,
          "g": 1,
          "r": 0.10588235294117647,
        },
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "backgroundColor": "#1bff00",
    }
  }
/>
`;

exports[`<Rectangle /> Rectangle with image background 1`] = `
<rectangle
  fills={
    Array [
      Object {
        "color": Object {
          "b": 0.1607843137254902,
          "g": 0.5450980392156862,
          "r": 1,
        },
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "backgroundColor": "#ff8b29",
      "backgroundImage": "./image.png",
      "backgroundSize": "cover",
    }
  }
/>
`;

exports[`<Rectangle /> Rectangle with layout style properties 1`] = `
<rectangle
  height={200}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "height": 200,
      "left": 20,
      "top": 50,
      "width": 100,
    }
  }
  width={100}
  x={20}
  y={50}
/>
`;

exports[`<Rectangle /> Rectangle without props 1`] = `
<rectangle
  innerRef={
    Object {
      "current": undefined,
    }
  }
/>
`;
